src/lib.rs

```diff
@@ -1042,6 +1059,75 @@ impl<const N: usize, T> CircularBuffer<N, T> {
         }
     }
 
+    /// Appends an element to the back of the buffer.
+    ///
+    /// # Panics
+    ///
+    /// If the buffer is full, the method will panic.
+    ///
+    /// # Examples
+    ///
+    /// ```should_panic
+    /// use circular_buffer::CircularBuffer;
+    ///
+    /// let mut buf = CircularBuffer::<3, char>::new();
+    ///
+    /// buf.push_back_or_panic('a'); assert_eq!(buf, ['a']);
+    /// buf.push_back_or_panic('b'); assert_eq!(buf, ['a', 'b']);
+    /// buf.push_back_or_panic('c'); assert_eq!(buf, ['a', 'b', 'c']);
+    /// // The buffer is now full; adding more values causes panic
+    /// buf.push_back_or_panic('d');
+    /// ```
+    pub fn push_back_or_panic(&mut self, item: T) {
+        if N == 0 {
+            // Nothing to do
+            return;
+        }
+        if self.size >= N {
+            // At capacity; panic
+            self.panic_on_full("push_back");
+        } else {
+            // Some uninitialized slots left; append at the end
+            self.inc_size();
+            self.back_maybe_uninit_mut().write(item);
+        }
+    }
+
+    /// Appends an element to the back of the buffer.
+    ///
+    /// # Errors
+    ///
+    /// If the buffer is full, the method will return the pushed value as an error.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use circular_buffer::CircularBuffer;
+    ///
+    /// let mut buf = CircularBuffer::<3, char>::new();
+    ///
+    /// buf.try_push_back('a'); assert_eq!(buf, ['a']);
+    /// buf.try_push_back('b'); assert_eq!(buf, ['a', 'b']);
+    /// buf.try_push_back('c'); assert_eq!(buf, ['a', 'b', 'c']);
+    /// // The buffer is now full; adding more values causes error
+    /// assert_eq!(buf.try_push_back('d').unwrap_err(), 'd')
+    /// ```
+    pub fn try_push_back(&mut self, item: T) -> Result<(), T> {
+        if N == 0 {
+            // Nothing to do
+            return Ok(());
+        }
+        if self.size >= N {
+            // At capacity; return the pushed item as error
+            Err(item)
+        } else {
+            // Some uninitialized slots left; append at the end
+            self.inc_size();
+            self.back_maybe_uninit_mut().write(item);
```

Comment:
Nit: I would just call `self.push_back(item)` here instead of duplicating the same logic. I believe the Rust compiler/LLVM are smart enough to optimize-out the duplicated size checks (but if you believe that's not the case, I'm happy to be proven wrong).

Reply:
I see two 'cmp' operations on the simplified [example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:24,endLineNumber:4,positionColumn:24,positionLineNumber:4,selectionStartColumn:24,selectionStartLineNumber:4,startColumn:24,startLineNumber:4),source:'const+N:+usize+%3D+10%3B%0A%0Apub+fn+push(arr:+%26mut+%5Bu32%3B+N%5D,+last_index:+usize,+element:+u32)+%7B%0A++++if+last_index+%3E%3DN+%7B%0A++++++++println!!(%22overwriting%22)%3B%0A++++%7D%0A++++else+%7B%0A++++++++arr%5Blast_index%5D+%3D+element%3B%0A++++%7D%0A%7D%0Apub+fn+try_push_back(arr:+%26mut%5Bu32%3B+N%5D,+last_index:+usize,+element:+u32)+-%3E+Result%3C(),+u32%3E+%7B%0A++++if+last_index+%3E%3DN+%7B%0A++++++++Err(element)%0A++++%7D+else+%7B%0A++++++++push(arr,+last_index,+element)%3B%0A++++++++Ok(())%0A++++%7D%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1720,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D2',overrides:!((name:edition,value:'2021')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.72.0+(Editor+%231)',t:'0')),header:(),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+1.66.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.72.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)